<launch>
	<arg name="veh" doc="Name of vehicle. ex: megaman"/>
	<arg name="local" default="false" doc="true for running on laptop. false for running on vehicle."/>
	<arg name="caffe_model" default="bvlc_trailnet"/>
	<arg name="config" default="baseline" doc="Specify a config."/>
    <arg name="param_file_name" default="default" doc="Specify a param file. ex:megaman." />
    <arg name="joy_mapper_param_file_name" default="$(arg param_file_name)" doc="Specify a joy_mapper param file. ex:high_speed" />
    <arg name="live" default="true" doc="live set to false means we don't run the actual camera (i.e. running from a log file" />

    # camera 
    <!-- camera_node -->
    <group if="$(arg live)">
        <include  file="$(find deep_lane_following)/launch/camera_white_balance.launch">
            <arg name="veh" value="$(arg veh)"/>
            <arg name="config" value="$(arg config)"/>
            <arg name="param_file_name" value="$(arg param_file_name)"/>
        </include>
    </group>

    #joystick
    <!-- joy -->
    <node ns="$(arg veh)" if="$(arg local)" pkg="joy" type="joy_node" name="joy" output="screen">
        <rosparam command="load" file="$(find duckietown)/config/$(arg config)/joy/joy_node/$(arg param_file_name).yaml"/>
    </node>
    <node ns="$(arg veh)" unless="$(arg local)" machine="$(arg veh)" pkg="joy" type="joy_node" name="joy" output="screen">
        <rosparam command="load" file="$(find duckietown)/config/$(arg config)/joy/joy_node/$(arg param_file_name).yaml"/>
    </node>

    #joy mapper
    <!-- joy_mapper -->
    <include file="$(find deep_lane_following)/launch/deep_joy_mapper.launch">
        <arg name="veh" value="$(arg veh)"/>
        <arg name="local" value="$(arg local)"/>
        <arg name="config" value="$(arg config)"/>
        <arg name="param_file_name" value="$(arg joy_mapper_param_file_name)"/>
    </include>

    #NCS prediction
    <remap from="ncs_caffe_prediction_node/compressed" to="camera_white_balance_node/image/compressed"/>
    <include file="$(find deep_lane_following)/launch/ncs_caffe_prediction.launch">
    	<arg name="veh" value="$(arg veh)"/>
    	<arg name="caffe_model" value="$(arg caffe_model)"/>
    	<arg name="local" value="$(arg local)"/>
        <arg name="config" value="$(arg config)"/>
        <arg name="param_file_name" value="$(arg param_file_name)"/>
    </include>	


    <!-- control -->
	<remap from="control_node/prediction_car_cmd" to="ncs_caffe_prediction_node/carcmd"/>
	<remap from="control_node/joy_car_cmd" to="joy_mapper_node/car_cmd"/>	
	<remap from="control_node/joystick_override" to="joy_mapper_node/joystick_override"/>
  	<include file="$(find deep_lane_following)/launch/control.launch">
    	<arg name="veh" value="$(arg veh)"/>
    	<arg name="config" value="$(arg config)"/>
  	</include>

    #car command
    <!-- car_cmd_switch_node -->
	<remap from="inverse_kinematics_node/car_cmd" to="control_node/car_cmd"/>
    <remap from="inverse_kinematics_node/wheels_cmd" to="wheels_driver_node/wheels_cmd" />
	<include file="$(find dagu_car)/launch/inverse_kinematics_node.launch">
		<arg name="veh" value="$(arg veh)"/>
		<arg name="local" value="$(arg local)"/>
		<arg name="config" value="$(arg config)"/>
	</include>

    #wheel driver
    <!-- Wheels Driver -->
    <include file="$(find dagu_car)/launch/wheels_driver_node.launch">
        <arg name="veh" value="$(arg veh)"/>
    </include>  



</launch>